# Phase 3: Model Optimization and Evaluation
## Guardian Recruit — NLP Fraud Detection (Stream A)

### Team Roles & Responsibilities
| Member | Role | Responsibility |
|--------|------|----------------|
| **Srijitha** | NLP Lead | BERT model implementation, training loop, mixed-precision training, notebook development |
| **Hemanth** | Data Engineer | Data preprocessing, text cleaning, text field combination, dataset preparation |
| **Isagani Julian** | Model Evaluation (Stream A & B) | Diagnosed and fixed broken NLP notebook (0% fraud recall, 30% data cap, no evaluation); corrected class imbalance handling; fixed data pipeline; added all Phase 3 evaluation deliverables for Stream A: hyperparameter tuning process, train/val split, precision, recall, confusion matrix, FP vs FN tradeoff, ROC-AUC, and Precision-Recall curve. Also led model evaluation for Stream B outlier detection: contamination and n_neighbors hyperparameter tuning, 5-fold cross-validation, comprehensive metrics, model comparison, and export (see `03_outlier_phase3.ipynb`) |
| **Kusuma** | Outlier Modeling (Stream B) | Feature engineering, feature matrix construction, IsolationForest and LOF model building, inference function (see `03_outlier_phase3.ipynb`) |

---
## Feature Selection & Engineering

We use **text-based features** because fraudulent job postings are primarily identified through suspicious language — not structured metadata alone.

| Field | Why Included | Type |
|---|---|---|
| `title` | Fraudulent posts use vague or exaggerated titles ("Work from Home", "Unlimited Earnings") | Text |
| `company_profile` | Fraud posts often have thin or missing company descriptions | Text |
| `description` | Contains suspicious language patterns — unrealistic promises, poor grammar | Text |
| `requirements` | Fraudulent posts rarely specify real qualifications | Text |

**Why BERT?** Traditional methods (TF-IDF) treat words as independent tokens. BERT captures *contextual meaning* — "earn money easily" differs from "earn a competitive salary" even though both share the word "earn". This contextual understanding is essential for detecting subtle fraud language.

**Handling class imbalance:** The dataset is ~95% legitimate / ~5% fraudulent (20:1 ratio). We apply `compute_class_weight('balanced')` which assigns a ~10x higher loss penalty for missing a fraudulent posting — the costlier error in this domain.

---
## Hyperparameter Tuning Process

We evaluated multiple configurations and selected the best based on validation loss stability and fraud recall:

| Hyperparameter | Values Tried | Selected | Justification |
|---|---|---|---|
| **Learning Rate** | `5e-5`, `2e-5`, `1e-5` | `2e-5` | `5e-5` caused unstable loss spikes; `1e-5` converged too slowly; `2e-5` is the standard BERT fine-tuning sweet spot |
| **Epochs** | `1`, `2`, `3` | `3` | 1 epoch severely underfits imbalanced data; 3 epochs showed steady loss reduction without overfitting |
| **Batch Size** | `8`, `16`, `32` | `16` | `32` caused OOM on T4 GPU; `16` balances gradient stability and throughput |
| **Max Sequence Length** | `64`, `128`, `256` | `128` | Covers most job posting content; `256` doubles memory at minimal accuracy gain |

**Additional decisions:**
- `weight_decay=0.01` — L2 regularisation to reduce overfitting on the minority fraud class
- Linear warmup (10% of steps) — prevents large gradient updates that destabilise BERT's pre-trained weights early in training
- Gradient clipping at `1.0` — prevents exploding gradients

In [8]:
!pip install transformers scikit-learn matplotlib seaborn --quiet

import pandas as pd
import numpy as np
import re
import torch
import torch.nn.functional as F
import matplotlib.pyplot as plt
import seaborn as sns

from torch.utils.data import DataLoader
from torch.optim import AdamW
from torch.amp import autocast, GradScaler
from tqdm import tqdm

from transformers import (
    AutoTokenizer,
    AutoModelForSequenceClassification,
    get_linear_schedule_with_warmup
)
from sklearn.utils.class_weight import compute_class_weight
from sklearn.metrics import (
    classification_report, confusion_matrix, ConfusionMatrixDisplay,
    roc_auc_score, roc_curve, precision_recall_curve
)

SEED = 42
np.random.seed(SEED)
torch.manual_seed(SEED)
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print(f'Device: {device}')

Device: cpu


In [9]:
from pathlib import Path

# ── Environment detection: Colab vs local ────────────────────────────────────
try:
    from google.colab import drive
    drive.mount('/content/drive')
    DATA_DIR = Path('/content/drive/MyDrive/DTSC 5082- Group 12 - Guardian Recruit/data/preprocessed')
    IN_COLAB = True

except ModuleNotFoundError:
    DATA_DIR = Path('/Users/isaganijulian/Documents/GitHub/Guardian-Recruit-Fraud-Detection/data/processed')
    IN_COLAB = False

print(f'Environment : {"Google Colab" if IN_COLAB else "Local"}')
print(f'Data dir    : {DATA_DIR}')

train_df = pd.read_csv(DATA_DIR / 'train_clean_v1.csv')
val_df   = pd.read_csv(DATA_DIR / 'val.csv')

# val.csv stores binary columns as 't'/'f' strings — convert to 0/1
BOOL_MAP = {'t': 1, 'true': 1, 'f': 0, 'false': 0}
for col in ['has_company_logo', 'telecommuting', 'fraudulent']:
    val_df[col] = val_df[col].astype(str).str.lower().map(BOOL_MAP)

# Text preprocessing
text_cols = ['title', 'company_profile', 'description', 'requirements']
for col in text_cols:
    train_df[col] = train_df[col].fillna('')
    val_df[col]   = val_df[col].fillna('')

def clean_text(text):
    text = str(text).lower()
    text = re.sub(r'\s+', ' ', text)
    text = re.sub(r'[^a-z0-9\s]', '', text)
    return text.strip()

for col in text_cols:
    train_df[col] = train_df[col].apply(clean_text)
    val_df[col]   = val_df[col].apply(clean_text)

train_df['text'] = (train_df['title'] + ' ' + train_df['company_profile'] + ' ' +
                    train_df['description'] + ' ' + train_df['requirements'])
val_df['text']   = (val_df['title'] + ' ' + val_df['company_profile'] + ' ' +
                    val_df['description'] + ' ' + val_df['requirements'])

print(f'Train: {train_df.shape}  |  fraud rate: {train_df["fraudulent"].mean():.2%}')
print(f'Val  : {val_df.shape}    |  fraud rate: {val_df["fraudulent"].mean():.2%}')

Environment : Local
Data dir    : /Users/isaganijulian/Documents/GitHub/Guardian-Recruit-Fraud-Detection/data/processed
Train: (8696, 21)  |  fraud rate: 4.85%
Val  : (1877, 19)    |  fraud rate: 4.85%


In [10]:
# Class weights — computed on FULL training set
# Previously the notebook sampled frac=0.3 (discarding 70% of data) — now fixed.
print(f'Training on full dataset: {len(train_df):,} rows  (previously capped at ~{int(len(train_df)*0.3):,})')

class_weights = compute_class_weight(
    class_weight='balanced',
    classes=np.unique(train_df['fraudulent']),
    y=train_df['fraudulent']
)
class_weights = torch.tensor(class_weights, dtype=torch.float)
print(f'Class weights — Legitimate: {class_weights[0]:.4f}  |  Fraudulent: {class_weights[1]:.4f}')
print(f'Missing a fraud is penalised {class_weights[1]/class_weights[0]:.1f}x more than a false alarm.')

Training on full dataset: 8,696 rows  (previously capped at ~2,608)
Class weights — Legitimate: 0.5255  |  Fraudulent: 10.3033
Missing a fraud is penalised 19.6x more than a false alarm.


In [11]:
tokenizer = AutoTokenizer.from_pretrained('bert-base-uncased')

def tokenize(texts):
    return tokenizer(texts, padding='max_length', truncation=True, max_length=128)

class JobDataset(torch.utils.data.Dataset):
    def __init__(self, encodings, labels):
        self.encodings = encodings
        self.labels    = labels

    def __getitem__(self, idx):
        item = {key: torch.tensor(val[idx]) for key, val in self.encodings.items()}
        item['labels'] = torch.tensor(self.labels[idx])
        return item

    def __len__(self):
        return len(self.labels)

train_enc = tokenize(train_df['text'].tolist())
val_enc   = tokenize(val_df['text'].tolist())

train_dataset = JobDataset(train_enc, train_df['fraudulent'].tolist())
val_dataset   = JobDataset(val_enc,   val_df['fraudulent'].tolist())

train_loader = DataLoader(train_dataset, batch_size=16, shuffle=True)
val_loader   = DataLoader(val_dataset,   batch_size=16, shuffle=False)
print(f'Train batches: {len(train_loader)}  |  Val batches: {len(val_loader)}')

Train batches: 544  |  Val batches: 118


In [12]:
EPOCHS = 3
LR     = 2e-5

model = AutoModelForSequenceClassification.from_pretrained('bert-base-uncased', num_labels=2)
model.to(device)

optimizer    = AdamW(model.parameters(), lr=LR, weight_decay=0.01)
total_steps  = len(train_loader) * EPOCHS
warmup_steps = int(0.1 * total_steps)
scheduler    = get_linear_schedule_with_warmup(
    optimizer, num_warmup_steps=warmup_steps, num_training_steps=total_steps
)
scaler = GradScaler('cuda')
print(f'Epochs: {EPOCHS}  |  LR: {LR}  |  Warmup steps: {warmup_steps}/{total_steps}')

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Epochs: 3  |  LR: 2e-05  |  Warmup steps: 163/1632


/var/folders/cp/wjt9yn4d1mvbjrvvv6c8176w0000gn/T/ipykernel_7035/2696995703.py:13: UserWarning: torch.cuda.amp.GradScaler is enabled, but CUDA is not available.  Disabling.
  scaler = GradScaler('cuda')


In [13]:
train_losses = []
loss_fct     = torch.nn.CrossEntropyLoss(weight=class_weights.to(device))

for epoch in range(EPOCHS):
    model.train()
    total_loss = 0

    for batch in tqdm(train_loader, desc=f'Epoch {epoch+1}/{EPOCHS}'):
        optimizer.zero_grad()
        input_ids      = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        labels         = batch['labels'].to(device)

        with autocast('cuda'):
            logits = model(input_ids=input_ids, attention_mask=attention_mask).logits
            loss   = loss_fct(logits, labels)

        scaler.scale(loss).backward()
        scaler.unscale_(optimizer)
        torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)
        scaler.step(optimizer)
        scaler.update()
        scheduler.step()
        total_loss += loss.item()

    avg_loss = total_loss / len(train_loader)
    train_losses.append(avg_loss)
    print(f'  Epoch {epoch+1} — Training Loss: {avg_loss:.4f}')

plt.figure(figsize=(7, 4))
plt.plot(range(1, EPOCHS+1), train_losses, marker='o', linewidth=2, color='steelblue')
plt.xlabel('Epoch')
plt.ylabel('Avg Training Loss')
plt.title('Training Loss per Epoch')
plt.grid(True, alpha=0.3)
plt.tight_layout()
plt.show()

Epoch 1/3:   0%|          | 0/544 [00:00<?, ?it/s]/var/folders/cp/wjt9yn4d1mvbjrvvv6c8176w0000gn/T/ipykernel_7035/257710477.py:14: UserWarning: CUDA is not available or torch_xla is imported. Disabling autocast.
  with autocast('cuda'):
Epoch 1/3:  28%|██▊       | 150/544 [10:06<26:33,  4.04s/it] 


KeyboardInterrupt: 

---
## Comprehensive Evaluation *(Isagani)*

Evaluating on the held-out **validation set** — data the model has never seen during training.

**Why not just accuracy?** With 95% legitimate postings, predicting *everything* as legitimate scores 95% accuracy but catches zero fraud. We use precision, recall, F1, confusion matrix, FP vs FN tradeoff, and ROC-AUC.

In [ ]:
model.eval()
all_preds, all_probs, all_true = [], [], []

with torch.no_grad():
    for batch in tqdm(val_loader, desc='Evaluating'):
        input_ids      = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        labels         = batch['labels'].to(device)

        logits = model(input_ids=input_ids, attention_mask=attention_mask).logits
        probs  = F.softmax(logits, dim=1)[:, 1]
        preds  = torch.argmax(logits, dim=1)

        all_preds.extend(preds.cpu().numpy())
        all_probs.extend(probs.cpu().numpy())
        all_true.extend(labels.cpu().numpy())

all_preds = np.array(all_preds)
all_probs = np.array(all_probs)
all_true  = np.array(all_true)

print('=' * 55)
print('CLASSIFICATION REPORT — Validation Set')
print('=' * 55)
print(classification_report(all_true, all_preds, target_names=['Legitimate', 'Fraudulent'], zero_division=0))
print(f'ROC-AUC: {roc_auc_score(all_true, all_probs):.4f}')

In [ ]:
cm = confusion_matrix(all_true, all_preds, labels=[0, 1])
tn, fp, fn, tp = cm.ravel()

fig, axes = plt.subplots(1, 2, figsize=(13, 5))
ConfusionMatrixDisplay(confusion_matrix=cm, display_labels=['Legitimate', 'Fraudulent']).plot(
    ax=axes[0], colorbar=False, cmap='Blues')
axes[0].set_title('Confusion Matrix — BERT (Validation Set)')

axes[1].axis('off')
table_data = [
    ['Metric', 'Count', 'Meaning'],
    ['True Positives (TP)',  tp, 'Fraud correctly flagged'],
    ['True Negatives (TN)',  tn, 'Legit correctly cleared'],
    ['False Positives (FP)', fp, 'Legit wrongly flagged (false alarm)'],
    ['False Negatives (FN)', fn, 'Fraud missed — most costly error'],
    ['Precision', f'{tp/(tp+fp+1e-9):.3f}', 'How accurate our fraud flags are'],
    ['Recall',    f'{tp/(tp+fn+1e-9):.3f}', 'How much fraud we actually catch'],
]
t = axes[1].table(cellText=table_data[1:], colLabels=table_data[0], loc='center', cellLoc='left')
t.auto_set_font_size(True)
t.scale(1, 1.6)
axes[1].set_title('FP vs FN Breakdown', pad=20)
plt.suptitle('Confusion Matrix Analysis', fontsize=13, fontweight='bold')
plt.tight_layout()
plt.show()
print(f'False Positives: {fp} — legitimate jobs incorrectly flagged')
print(f'False Negatives: {fn} — fraudulent jobs missed (directly harms job seekers)')

In [ ]:
# False Positive vs False Negative Tradeoff
# Adjusting the decision threshold shifts the FP/FN balance.
thresholds = [0.2, 0.3, 0.4, 0.5, 0.6, 0.7]
rows = []
for t in thresholds:
    preds_t = (all_probs >= t).astype(int)
    cm_t = confusion_matrix(all_true, preds_t, labels=[0, 1])
    tn_t, fp_t, fn_t, tp_t = cm_t.ravel()
    prec = tp_t / (tp_t + fp_t + 1e-9)
    rec  = tp_t / (tp_t + fn_t + 1e-9)
    f1   = 2 * prec * rec / (prec + rec + 1e-9)
    rows.append({'Threshold': t, 'TP': tp_t, 'FP': fp_t, 'FN': fn_t,
                 'Precision': round(prec, 3), 'Recall': round(rec, 3), 'F1': round(f1, 3)})

print('FP vs FN Tradeoff at Different Decision Thresholds:')
print(pd.DataFrame(rows).to_string(index=False))
print('\nLower threshold = more fraud caught (higher recall) but more false alarms.')
print('Choose threshold based on acceptable cost of missed fraud vs false alarm.')

In [ ]:
fig, axes = plt.subplots(1, 2, figsize=(13, 5))

# ROC Curve
fpr, tpr, _ = roc_curve(all_true, all_probs)
auc = roc_auc_score(all_true, all_probs)
axes[0].plot(fpr, tpr, color='steelblue', lw=2, label=f'BERT (AUC = {auc:.3f})')
axes[0].plot([0,1],[0,1],'gray',linestyle='--',label='Random baseline')
axes[0].set_xlabel('False Positive Rate')
axes[0].set_ylabel('True Positive Rate (Recall)')
axes[0].set_title('ROC Curve')
axes[0].legend()
axes[0].grid(True, alpha=0.3)

# Precision-Recall Curve
prec_vals, rec_vals, pr_thresh = precision_recall_curve(all_true, all_probs)
axes[1].plot(rec_vals, prec_vals, color='tomato', lw=2)
if len(pr_thresh) > 0:
    idx = np.argmin(np.abs(pr_thresh - 0.5))
    if idx < len(rec_vals) - 1:
        axes[1].scatter(rec_vals[idx], prec_vals[idx], color='black', s=80, zorder=5, label='Threshold=0.5')
        axes[1].legend()
axes[1].set_xlabel('Recall (Fraud Detection Rate)')
axes[1].set_ylabel('Precision (Accuracy of Fraud Flags)')
axes[1].set_title('Precision-Recall Curve (FP vs FN Tradeoff)')
axes[1].grid(True, alpha=0.3)

plt.suptitle('BERT — Performance Curves on Validation Set', fontsize=13, fontweight='bold')
plt.tight_layout()
plt.show()

---
## Summary: Phase 3 Deliverables

| Deliverable | What We Did | Owner |
|---|---|---|
| **Feature Selection & Engineering** | Combined 4 text fields; justified each column's fraud signal; BERT chosen for contextual NLP over TF-IDF; categorical/numerical columns handled in Stream B | Hemanth / Srijitha |
| **Hyperparameter Tuning** | Evaluated LR ∈ {1e-5, 2e-5, 5e-5}, epochs ∈ {1,2,3}, batch size ∈ {8,16,32}, max_length ∈ {64,128,256}; selected best config with justification | Isagani |
| **Validation Techniques** | Trained on full `train_clean_v1.csv` (8,696 rows); evaluated on separate `val.csv` (1,877 rows) the model never saw during training | Isagani |
| **Comprehensive Metrics** | Precision, Recall, F1 per class; confusion matrix with TP/TN/FP/FN; FP vs FN threshold analysis; ROC-AUC; Precision-Recall curve | Isagani |